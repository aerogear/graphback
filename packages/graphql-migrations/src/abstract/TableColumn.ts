export type TableColumnType =
  'integer' |
  'bigInteger' |
  'text' |
  'string' |
  'float' |
  'decimal' |
  'boolean' |
  'date' |
  'datetime' |
  'time' |
  'timestamp' |
  'binary' |
  'enum' |
  'json' |
  'jsonb' |
  'uuid'

export interface ForeignKey {
  type: string | undefined
  field: string | undefined
  tableName: string | undefined
  columnName: string | undefined
}

export interface TableColumn {
  name: string
  comment: string | undefined
  annotations: any
  type: string
  args: any[]
  nullable: boolean
  foreign: ForeignKey | undefined
  defaultValue: any,
  isPrimaryKey: boolean
  autoIncrementable: boolean
}
