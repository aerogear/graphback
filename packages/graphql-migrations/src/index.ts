export { migrateDB, MigrateOptions } from './migrate'
export { generateAbstractDatabase } from './abstract/generateAbstractDatabase'
export { computeDiff } from './diff/computeDiff'
export { read } from './connector/read'
export { write } from './connector/write'
export { MigratePlugin, WriteParams } from './plugin/MigratePlugin'
export * from './plugin/MigrateOperationFilter'

