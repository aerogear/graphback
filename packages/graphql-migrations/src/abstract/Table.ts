import { TableColumn } from './TableColumn'

export interface Table {
  name: string
  comment: string | undefined
  annotations: any
  columns: TableColumn[]
  columnMap: Map<string, TableColumn>
  indexes: TableIndex[]
  uniques: TableUnique[]
}

export interface TableIndex {
  columns: string[]
  name: string | undefined
  type: string | undefined
}

export interface TableUnique {
  columns: string[]
  name: string | undefined
}
