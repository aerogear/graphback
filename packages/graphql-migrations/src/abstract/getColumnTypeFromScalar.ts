import { GraphQLField, GraphQLScalarType } from 'graphql'
import { parseAnnotationsCompat } from '../util/parseAnnotationsCompat';
import { TableColumnType } from './TableColumn'

export interface TableColumnTypeDescriptor {
  /**
   * Knex column builder function name.
   */
  type: TableColumnType | string
  /**
   * Builder function arguments.
   */
  args: any[]
}

// eslint-disable-next-line complexity
// eslint-disable-next-line
export default function(
  field: GraphQLField<any, any>,
  scalarType: GraphQLScalarType,
  annotations: any,
): TableColumnTypeDescriptor | null {
  if (!annotations) {
    annotations = parseAnnotationsCompat('db', field.description || undefined)
  }

  // increments
  if (scalarType && scalarType.name === 'ID') {
    if (annotations.type) {
      throw new Error(`@db(type: <type>) annotation is not permitted on ID field.`);
    }

    return {
      type: 'increments',
      args: [],
    }
  }


  // text
  if (annotations.type === 'text') {
    return {
      type: 'text',
      args: [],
    }
  }

  // string
  if ((scalarType && scalarType.name === 'String')|| annotations.type === 'string') {
    return {
      type: 'string',
      args: [annotations.length || 255],
    }
  }

  // integer
  if ((scalarType && scalarType.name === 'Int') || annotations.type === 'integer') {
    return {
      type: 'integer',
      args: [],
    }
  }

  // float
  if ((scalarType && scalarType.name === 'Float') || annotations.type === 'float') {
    return {
      type: 'float',
      args: [annotations.precision, annotations.scale],
    }
  }

  // boolean
  if ((scalarType && scalarType.name === 'Boolean') || annotations.type === 'boolean') {
    return {
      type: 'boolean',
      args: [],
    }
  }

  // date
  if (annotations.type === 'date' || (scalarType && scalarType.name === 'GraphbackDate')) {
    return {
      type: 'date',
      args: [],
    }
  }

  // datetime & time
  if (['datetime', 'time'].includes(annotations.type)) {
    return {
      type: annotations.type,
      args: [annotations.precision],
    }
  }

  // graphback datetime
  if (scalarType && scalarType.name === 'GraphbackDateTime') {
    return {
      type: "datetime",
      args: [],
    }
  }

  // graphback time
  if (scalarType && scalarType.name === 'GraphbackTime') {
    return {
      type: "time",
      args: [],
    }
  }

  // timestamp
  if (annotations.type === 'timestamp') {
    return {
      type: 'timestamp',
      args: [annotations.useTz, annotations.precision],
    }
  }

  // graphback timestamp
  if (scalarType && scalarType.name === 'GraphbackTimestamp') {
    return {
      type: "timestamp",
      args: [],
    }
  }

  // binary
  if (annotations.type === 'binary') {
    return {
      type: 'binary',
      args: [annotations.length],
    }
  }

  // json & jsonb
  if (['json', 'jsonb'].includes(annotations.type)) {
    return {
      type: annotations.type,
      args: [],
    }
  }

  // graphback json
  if (scalarType && (scalarType.name === 'GraphbackJSON' || scalarType.name === 'GraphbackJSONObject')) {
    return {
      type: "json",
      args: [],
    }
  }


  // graphback ObjectId
  if (scalarType && scalarType.name === 'GraphbackObjectID') {
    return {
      type: "string",
      args: [24]
    }
  }

  // unknown type
  if (annotations.type) {
    return {
      type: annotations.type,
      args: annotations.args || [],
    }
  }

  return undefined
}
