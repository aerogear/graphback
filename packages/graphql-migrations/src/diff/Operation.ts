import { AbstractDatabase } from '../abstract/AbstractDatabase';

export type OperationType =
  'table.create' |
  'table.rename' |
  'table.comment.set' |
  'table.drop' |
  'table.index.create' |
  'table.index.drop' |
  'table.primary.set' |
  'table.primary.drop' |
  'table.unique.create' |
  'table.unique.drop' |
  'table.foreign.create' |
  'table.foreign.drop' |
  'column.create' |
  'column.rename' |
  'column.alter' |
  'column.drop'

export interface Operation {
  type: OperationType
  priority: number
}

export interface MigrationResults {
  results: Operation[]
  previousDB: AbstractDatabase
  newDB: AbstractDatabase
}

export interface TableCreateOperation extends Operation {
  type: 'table.create'
  table: string
}

export interface TableRenameOperation extends Operation {
  type: 'table.rename'
  fromName: string
  toName: string
}

export interface TableCommentSetOperation extends Operation {
  type: 'table.comment.set'
  table: string
  comment: string | undefined
}

export interface TableDropOperation extends Operation {
  type: 'table.drop'
  table: string
}

export interface TableIndexCreateOperation extends Operation {
  type: 'table.index.create'
  table: string
  columns: string[]
  indexName: string | undefined
  indexType: string | undefined
}

export interface TableIndexDropOperation extends Operation {
  type: 'table.index.drop'
  table: string
  columns: string[]
  indexName: string | undefined
}

export interface TablePrimarySetOperation extends Operation {
  type: 'table.primary.set'
  table: string
  column: string
  columnType: string
}

export interface TablePrimaryDropOperation extends Operation {
  type: 'table.primary.drop'
  table: string,
  primaryKeyName: string
}

export interface TableUniqueCreateOperation extends Operation {
  type: 'table.unique.create'
  table: string
  columns: string[]
  indexName: string | undefined
}

export interface TableUniqueDropOperation extends Operation {
  type: 'table.unique.drop'
  table: string
  columns: string[]
  indexName: string | undefined
}

export interface TableForeignCreateOperation extends Operation {
  type: 'table.foreign.create'
  table: string
  column: string
  referenceTable: string
  referenceColumn: string
}

export interface TableForeignDropOperation extends Operation {
  type: 'table.foreign.drop'
  table: string
  column: string
}

export interface ColumnCreateOperation extends Operation {
  type: 'column.create'
  table: string
  column: string
  columnType: string
  args: any[]
  comment: string | undefined
  nullable: boolean
  defaultValue: any
  isPrimaryKey: boolean
  autoIncrementable: boolean
}

export interface ColumnAlterOperation extends Operation {
  type: 'column.alter'
  table: string
  column: string
  columnType: string
  args: any[]
  comment: string | undefined
  nullable: boolean
  defaultValue: any,
  isPrimaryKey: boolean
}

export interface ColumnRenameOperation extends Operation {
  type: 'column.rename'
  table: string
  fromName: string
  toName: string
}

export interface ColumnDropOperation extends Operation {
  type: 'column.drop'
  table: string
  column: string
}
