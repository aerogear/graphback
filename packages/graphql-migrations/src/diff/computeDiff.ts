/*eslint-disable max-lines */
import { isEqual } from 'lodash'
import { AbstractDatabase } from '../abstract/AbstractDatabase'
import { Table, TableIndex, TableUnique } from '../abstract/Table'
import { TableColumn } from '../abstract/TableColumn'
import getKnexColumnType from '../util/getKnexColumnType'
import * as Operations from './Operation'

export async function computeDiff(from: AbstractDatabase, to: AbstractDatabase, {
  updateComments = false,
}: any = {}): Promise<Operations.Operation[]> {
  // eslint-disable-next-line @typescript-eslint/no-use-before-define
  const differ = new Differ(from, to, {
    updateComments,
  })

  return differ.diff()
}

// eslint-disable-next-line @typescript-eslint/tslint/config
class Differ {
  private from: AbstractDatabase
  private to: AbstractDatabase
  private updateComments: boolean
  private operations: Operations.Operation[] = []
  private tableCount: number = 0

  // eslint-disable-next-line @typescript-eslint/explicit-member-accessibility
  constructor(from: AbstractDatabase, to: AbstractDatabase, options: any) {
    this.from = from
    this.to = to
    this.updateComments = options.updateComments
  }

  // eslint-disable-next-line complexity
  public diff(): Operations.Operation[] {
    this.operations.length = 0

    // tslint:disable-next-line: array-type
    const sameTableQueue: { fromTable: Table, toTable: Table }[] = []
    const addTableQueue = this.to.tables.slice()
    for (const fromTable of this.from.tables) {
      let removed = true
      for (let i = 0, l = addTableQueue.length; i < l; i++) {
        const toTable = addTableQueue[i]
        // Same table
        if (toTable.name === fromTable.name) {
          removed = false
        }

        // Rename table
        const { annotations } = toTable
        if (annotations && annotations.oldNames && annotations.oldNames.includes(fromTable.name)) {
          removed = false
          this.renameTable(fromTable, toTable)
        }

        // Same or Rename
        if (!removed) {
          sameTableQueue.push({ fromTable, toTable })
          // A new table shouldn't be added
          addTableQueue.splice(i, 1)
          break
        }
      }

      // Drop table
      if (removed) {
        this.dropTable(fromTable)
      }
    }

    // Create table
    for (const toTable of addTableQueue) {
      this.createTable(toTable)
    }

    // Compare tables
    for (const { fromTable, toTable } of sameTableQueue) {
      // Comment
      if (this.updateComments && fromTable.comment !== toTable.comment) {
        this.setTableComment(toTable)
      }

      let dropPrimaryKey = false;

      const sameColumnQueue: { fromCol: TableColumn, toCol: TableColumn }[] = []
      const addColumnQueue = toTable.columns.slice()
      for (const fromCol of fromTable.columns) {
        let removed = true
        for (let i = 0, l = addColumnQueue.length; i < l; i++) {
          const toCol = addColumnQueue[i]
          // Same column
          if (toCol.name === fromCol.name) {
            removed = false
          }

          // Rename column
          const { annotations } = toCol
          if (annotations && annotations.oldNames && annotations.oldNames.includes(fromCol.name)) {
            removed = false
            this.renameColumn(toTable, fromCol, toCol)
          }

          // Same or Rename
          if (!removed) {
            sameColumnQueue.push({ fromCol, toCol })
            dropPrimaryKey = dropPrimaryKey || (fromCol.isPrimaryKey && !toCol.isPrimaryKey);
            // A new table shouldn't be added
            addColumnQueue.splice(i, 1)
            break
          }

          dropPrimaryKey = dropPrimaryKey || fromCol.isPrimaryKey;
        }

        // Drop column
        if (removed) {
          this.dropColumn(fromTable, fromCol)
        }
      }

      // Drop primary key
      if (dropPrimaryKey) {
        this.dropPrimary(toTable); 
      }

      // Add columns
      for (const column of addColumnQueue) {
        if (column.isPrimaryKey && column.autoIncrementable) {
          this.setPrimary(toTable, column)
        } else {
          this.createColumn(toTable, column)
        }
      }

      // Compare columns
      for (const { fromCol, toCol } of sameColumnQueue) {
        if (toCol.name === 'id') {
          const fromPrimaryKeyType = getKnexColumnType(fromCol.type);
          const toPrimaryKeyType = getKnexColumnType(toCol.type);

          if (fromPrimaryKeyType !== toPrimaryKeyType) {
            console.warn(`Cannot change type of column ${toTable.name}.${toCol.name} from ${fromCol.type} to ${toCol.type}`);
          }

          continue;
        }
        if ((this.updateComments && fromCol.comment !== toCol.comment) ||
          getKnexColumnType(fromCol.type) !== getKnexColumnType(toCol.type) || !isEqual(fromCol.args, toCol.args) ||
          fromCol.nullable !== toCol.nullable ||
          fromCol.defaultValue !== toCol.defaultValue ||
          fromCol.isPrimaryKey !== toCol.isPrimaryKey ||
          (Array.isArray(fromCol.defaultValue) && !isEqual(fromCol.defaultValue, toCol.defaultValue)) ||
          (typeof fromCol.defaultValue === 'object' && !isEqual(fromCol.defaultValue, toCol.defaultValue))
        ) {
          this.alterColumn(toTable, toCol)
        }

        // Foreign key
        if ((fromCol.foreign && !toCol.foreign) ||
          (!fromCol.foreign && toCol.foreign) ||
          (fromCol.foreign && toCol.foreign &&
            (fromCol.foreign.tableName !== toCol.foreign.tableName ||
              fromCol.foreign.columnName !== toCol.foreign.columnName)
          )) {
          if (fromCol.foreign) { this.dropForeignKey(toTable, fromCol) }
          if (toCol.foreign) { this.createForeignKey(toTable, toCol) }
        }
      }

      // Index
      this.compareIndex(
        fromTable.indexes,
        toTable.indexes,
        // @ts-ignore
        (index: TableIndex) => this.createIndex(toTable, index),
        (index: TableIndex) => this.dropIndex(fromTable, index),
      )

      // Unique contraint
      this.compareIndex(
        fromTable.uniques,
        toTable.uniques,
        (index: TableUnique) => this.createUnique(toTable, index),
        (index: TableUnique) => this.dropUnique(fromTable, index),
      )
    }

    return this.operations
  }

  private createTable(table: Table) {
    const op: Operations.TableCreateOperation = {
      type: 'table.create',
      table: table.name,
      priority: this.tableCount++,
    }
    this.operations.push(op)

    // Comment
    if (table.comment) {
      this.setTableComment(table)
    }

    // Columns
    for (const column of table.columns) {
      if (column.isPrimaryKey && column.autoIncrementable) {
        this.setPrimary(table, column)
      } else {
        this.createColumn(table, column);
      }
    }

    // Index
    for (const index of table.indexes) {
      this.createIndex(table, index)
    }

    // Unique contraint
    for (const index of table.uniques) {
      this.createUnique(table, index)
    }
  }

  private renameTable(fromTable: Table, toTable: Table) {
    const op: Operations.TableRenameOperation = {
      type: 'table.rename',
      fromName: fromTable.name,
      toName: toTable.name,
      priority: 0,
    }
    this.operations.push(op)
  }

  private dropTable(table: Table) {
    const op: Operations.TableDropOperation = {
      type: 'table.drop',
      table: table.name,
      priority: 0,
    }
    this.operations.push(op)
  }

  private setTableComment(table: Table) {
    const op: Operations.TableCommentSetOperation = {
      type: 'table.comment.set',
      table: table.name,
      comment: table.comment,
      priority: 0,
    }
    this.operations.push(op)
  }

  private setPrimary(table: Table, column: TableColumn) {
    const op: Operations.TablePrimarySetOperation = {
      type: 'table.primary.set',
      table: table.name,
      columnType: column.type,
      column: column.name,
      priority: 0,
    }
    this.operations.push(op)
  }


  private dropPrimary(table: Table) {
    const op: Operations.TablePrimaryDropOperation = {
      type: 'table.primary.drop',
      table: table.name,
      primaryKeyName: `${table.name}_pkey`, // TODO - this is the default primary key, it maybe be different we should correctly determine primary key name
      priority: 0,
    }
    this.operations.push(op)
  }

  private createIndex(table: Table, index: TableIndex) {
    const op: Operations.TableIndexCreateOperation = {
      type: 'table.index.create',
      table: table.name,
      columns: index.columns,
      indexName: index.name,
      indexType: index.type,
      priority: 0,
    }
    this.operations.push(op)
  }

  private dropIndex(table: Table, index: TableIndex) {
    const op: Operations.TableIndexDropOperation = {
      type: 'table.index.drop',
      table: table.name,
      columns: index.columns,
      indexName: index.name,
      priority: 0,
    }
    this.operations.push(op)
  }

  private createUnique(table: Table, index: TableUnique) {
    const op: Operations.TableUniqueCreateOperation = {
      type: 'table.unique.create',
      table: table.name,
      columns: index.columns,
      indexName: index.name,
      priority: 0,
    }
    this.operations.push(op)
  }

  /**
   * @param {Table} table
   * @param {TableUnique} index
   */
  private dropUnique(table: Table, index: TableUnique) {
    const op: Operations.TableUniqueDropOperation = {
      type: 'table.unique.drop',
      table: table.name,
      columns: index.columns,
      indexName: index.name,
      priority: 0,
    }
    this.operations.push(op)
  }

  private createForeignKey(table: Table, column: TableColumn) {
    if (column.foreign && column.foreign.tableName && column.foreign.columnName) {
      const op: Operations.TableForeignCreateOperation = {
        type: 'table.foreign.create',
        table: table.name,
        column: column.name,
        referenceTable: column.foreign.tableName,
        referenceColumn: column.foreign.columnName,
        priority: 0,
      }
      this.operations.push(op)
    }
  }

  private dropForeignKey(table: Table, column: TableColumn) {
    if (column.foreign) {
      const op: Operations.TableForeignDropOperation = {
        type: 'table.foreign.drop',
        table: table.name,
        column: column.name,
        priority: 0,
      }
      this.operations.push(op)
    }
  }

  private createColumn(table: Table, column: TableColumn) {
    const op: Operations.ColumnCreateOperation = {
      type: 'column.create',
      table: table.name,
      column: column.name,
      columnType: getKnexColumnType(column.type), // TODO: Move column type mapping to abstraction layer
      args: column.args,
      comment: column.comment,
      nullable: column.nullable,
      defaultValue: column.defaultValue,
      priority: 0,
      isPrimaryKey: column.isPrimaryKey,
      autoIncrementable: column.autoIncrementable
    }
    this.operations.push(op)

    // Foreign key
    this.createForeignKey(table, column)
  }

  private renameColumn(table: Table, fromCol: TableColumn, toCol: TableColumn) {
    const op: Operations.ColumnRenameOperation = {
      type: 'column.rename',
      table: table.name,
      fromName: fromCol.name,
      toName: toCol.name,
      priority: 0,
    }
    this.operations.push(op)
  }

  private alterColumn(table: Table, column: TableColumn) {
    const op: Operations.ColumnAlterOperation = {
      type: 'column.alter',
      table: table.name,
      column: column.name,
      columnType: column.type,
      args: column.args,
      comment: column.comment,
      nullable: column.nullable,
      defaultValue: column.defaultValue,
      priority: 0,
      isPrimaryKey: column.isPrimaryKey
    }
    this.operations.push(op)
  }

  private dropColumn(table: Table, column: TableColumn) {
    const op: Operations.ColumnDropOperation = {
      type: 'column.drop',
      table: table.name,
      column: column.name,
      priority: 0,
    }
    this.operations.push(op)
  }

  private compareIndex(
    fromList: (TableIndex | TableUnique)[],
    toList: (TableIndex | TableUnique)[],
    create: (index: TableIndex | TableUnique) => void,
    drop: (index: TableIndex | TableUnique) => void,
  ) {
    const addIndexQueue = toList.slice()
    for (const fromIndex of fromList) {
      let removed = true
      for (let i = 0, l = addIndexQueue.length; i < l; i++) {
        const toIndex = addIndexQueue[i]
        if (
          fromIndex.name === toIndex.name &&
          // @ts-ignore
          fromIndex.type === toIndex.type &&
          isEqual(fromIndex.columns.sort(), toIndex.columns.sort())
        ) {
          removed = false
          addIndexQueue.splice(i, 1)
          break
        }
      }

      if (removed) {
        drop(fromIndex)
      }
    }
    for (const index of addIndexQueue) {
      create(index)
    }
  }
}
