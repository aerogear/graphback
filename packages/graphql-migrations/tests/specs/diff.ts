/*eslint-disable max-lines*/
import { computeDiff } from '../../src';
import { AbstractDatabase } from '../../src/abstract/AbstractDatabase'
import { Table } from '../../src/abstract/Table'
import { TableColumn } from '../../src/abstract/TableColumn'
import { Operation } from '../../src/diff/Operation';

function dbFactory(tables: Table[] = []): AbstractDatabase {
  return {
    tables,
    tableMap: new Map(),
  }
}

function tableFactory(options: any): Table {
  return {
    columns: [],
    indexes: [],
    uniques: [],
    annotations: {},
    columnMap: new Map(),
    ...options,
  }
}

function columnFactory(options: any): TableColumn {
  return {
    args: [],
    nullable: true,
    annotations: {},
    defaultValue: undefined,
    comment: undefined,
    foreign: undefined,
    ...options,
  }
}

test('create simple table', async () => {
  const result = await computeDiff(dbFactory(), dbFactory([
    tableFactory({
      name: 'User',
      comment: 'Some comment',
      columns: [
        columnFactory({
          name: 'id',
          type: 'increments',
          nullable: false,
          isPrimaryKey: true,
          autoIncrementable: true
        }),
        columnFactory({
          name: 'name',
          type: 'string',
          args: [150],
          isPrimaryKey: false,
          autoIncrementable: false
        }),
      ]
    }),
  ]))
  expect(result.length).toEqual(4);

  expect(result[0]).toEqual({
    type: 'table.create',
    table: 'User',
    priority: 0,
  } as Operation)
  expect(result[1]).toEqual({
    type: 'table.comment.set',
    table: 'User',
    comment: 'Some comment',
    priority: 0,
  } as Operation)
  expect(result[2]).toEqual({
    type: "table.primary.set",
    table: 'User',
    column: 'id',
    columnType: "increments",
    priority: 0,
  } as Operation)
  expect(result[3]).toEqual({
    type: 'column.create',
    table: 'User',
    column: 'name',
    columnType: 'string',
    args: [150],
    nullable: true,
    defaultValue: undefined,
    comment: undefined,
    isPrimaryKey: false,
    autoIncrementable: false,
    priority: 0,
  } as Operation)

  expect(result).toMatchSnapshot();
})

test('rename table', async () => {
  const result = await computeDiff(dbFactory([
    tableFactory({
      name: 'User',
    }),
  ]), dbFactory([
    tableFactory({
      name: 'users',
      annotations: {
        oldNames: ['User'],
      },
    }),
  ]))
  expect(result.length).toEqual(1);
  expect(result[0]).toEqual({
    type: 'table.rename',
    fromName: 'User',
    toName: 'users',
    priority: 0,
  } as Operation)
})

test('update table comment', async () => {
  const result = await computeDiff(dbFactory([
    tableFactory({
      name: 'User',
      comment: 'Some comment',
    }),
  ]), dbFactory([
    tableFactory({
      name: 'User',
      comment: 'New comment',
    }),
  ]), {
    updateComments: true,
  })
  expect(result.length).toEqual(1);
  expect(result[0]).toEqual({
    type: 'table.comment.set',
    table: 'User',
    comment: 'New comment',
    priority: 0,
  } as Operation)
})

test('set primary key', async () => {
  const result = await computeDiff(dbFactory([
    tableFactory({
      name: 'User'
    }),
  ]), dbFactory([
    tableFactory({
      name: 'User',
      columns: [
        columnFactory({
          name: 'id',
          type: 'increments',
          nullable: false,
          isPrimaryKey: true,
          autoIncrementable: true
        }),
      ],
    }),
  ]))
  expect(result.length).toEqual(1);
  expect(result[0]).toEqual({
    type: 'table.primary.set',
    table: 'User',
    column: 'id',
    columnType: 'increments',
    priority: 0,
  } as Operation)
})

test('add column', async () => {
  const result = await computeDiff(dbFactory([
    tableFactory({
      name: 'User',
      columns: [
        columnFactory({
          name: 'id',
          type: 'increments',
        })
      ],
    }),
  ]), dbFactory([
    tableFactory({
      name: 'User',
      columns: [
        columnFactory({
          name: 'id',
          type: 'increments',
        }),
        columnFactory({
          name: 'name',
          type: 'string',
        }),
      ],
    }),
  ]))
  expect(result.length).toEqual(1);
  expect(result[0]).toEqual({
    type: 'column.create',
    table: 'User',
    column: 'name',
    columnType: 'string',
    args: [],
    nullable: true,
    defaultValue: undefined,
    comment: undefined,
    priority: 0,
  } as Operation)
})

test('add and remove column', async () => {
  const result = await computeDiff(dbFactory([
    tableFactory({
      name: 'User',
      columns: [
        columnFactory({
          name: 'id',
          type: 'increments',
        }),
      ],
    }),
  ]), dbFactory([
    tableFactory({
      name: 'User',
      columns: [
        columnFactory({
          name: 'email',
          type: 'string',
        }),
      ],
    }),
  ]))
  expect(result.length).toEqual(2);
  expect(result[0]).toEqual({
    type: 'column.drop',
    table: 'User',
    column: 'id',
    priority: 0,
  } as Operation)
  expect(result[1]).toEqual({
    type: 'column.create',
    table: 'User',
    column: 'email',
    columnType: 'string',
    args: [],
    nullable: true,
    defaultValue: undefined,
    comment: undefined,
    priority: 0,
  } as Operation)
})

test('rename column', async () => {
  const result = await computeDiff(dbFactory([
    tableFactory({
      name: 'User',
      columns: [
        columnFactory({
          name: 'id',
          type: 'increments',
        }),
      ],
    }),
  ]), dbFactory([
    tableFactory({
      name: 'User',
      columns: [
        columnFactory({
          name: 'email',
          type: 'increments',
          annotations: {
            oldNames: ['id'],
          },
        }),
      ],
    }),
  ]))
  expect(result.length).toEqual(1);
  expect(result[0]).toEqual({
    type: 'column.rename',
    table: 'User',
    fromName: 'id',
    toName: 'email',
    priority: 0,
  } as Operation)
})

test('change column comment', async () => {
  const result = await computeDiff(dbFactory([
    tableFactory({
      name: 'User',
      columns: [
        columnFactory({
          name: 'name',
          type: 'string',
          comment: 'foo',
        }),
      ],
    }),
  ]), dbFactory([
    tableFactory({
      name: 'User',
      columns: [
        columnFactory({
          name: 'name',
          type: 'string',
          comment: 'bar',
        }),
      ],
    }),
  ]), {
    updateComments: true,
  })
  expect(result.length).toEqual(1);
  expect(result[0]).toEqual({
    type: 'column.alter',
    table: 'User',
    column: 'name',
    columnType: 'string',
    args: [],
    nullable: true,
    defaultValue: undefined,
    comment: 'bar',
    priority: 0,
  } as Operation)
})

test('change column type', async () => {
  const result = await computeDiff(dbFactory([
    tableFactory({
      name: 'User',
      columns: [
        columnFactory({
          name: 'age',
          type: 'integer',
        }),
      ],
    }),
  ]), dbFactory([
    tableFactory({
      name: 'User',
      columns: [
        columnFactory({
          name: 'age',
          type: 'bigInteger',
        }),
      ],
    }),
  ]))
  expect(result.length).toEqual(1);
  expect(result[0]).toEqual({
    type: 'column.alter',
    table: 'User',
    column: 'age',
    columnType: 'bigInteger',
    args: [],
    nullable: true,
    defaultValue: undefined,
    comment: undefined,
    priority: 0,
  } as Operation)
})

test('change column type args', async () => {
  const result = await computeDiff(dbFactory([
    tableFactory({
      name: 'User',
      columns: [
        columnFactory({
          name: 'name',
          type: 'string',
          args: [100],
        }),
      ],
    }),
  ]), dbFactory([
    tableFactory({
      name: 'User',
      columns: [
        columnFactory({
          name: 'name',
          type: 'string',
          args: [200],
        }),
      ],
    }),
  ]))
  expect(result.length).toEqual(1);
  expect(result[0]).toEqual({
    type: 'column.alter',
    table: 'User',
    column: 'name',
    columnType: 'string',
    args: [200],
    nullable: true,
    defaultValue: undefined,
    comment: undefined,
    priority: 0,
  } as Operation)
})

test('change column nullable', async () => {
  const result = await computeDiff(dbFactory([
    tableFactory({
      name: 'User',
      columns: [
        columnFactory({
          name: 'name',
          type: 'string',
          nullable: false,
        }),
      ],
    }),
  ]), dbFactory([
    tableFactory({
      name: 'User',
      columns: [
        columnFactory({
          name: 'name',
          type: 'string',
          nullable: true,
        }),
      ],
    }),
  ]))
  expect(result.length).toEqual(1);
  expect(result[0]).toEqual({
    type: 'column.alter',
    table: 'User',
    column: 'name',
    columnType: 'string',
    args: [],
    nullable: true,
    defaultValue: undefined,
    comment: undefined,
    priority: 0,
  } as Operation)
})

test('change column default value', async () => {
  const result = await computeDiff(dbFactory([
    tableFactory({
      name: 'User',
      columns: [
        columnFactory({
          name: 'name',
          type: 'string',
          defaultValue: 'foo',
        }),
      ],
    }),
  ]), dbFactory([
    tableFactory({
      name: 'User',
      columns: [
        columnFactory({
          name: 'name',
          type: 'string',
          defaultValue: 'bar',
        }),
      ],
    }),
  ]))
  expect(result.length).toEqual(1);
  expect(result[0]).toEqual({
    type: 'column.alter',
    table: 'User',
    column: 'name',
    columnType: 'string',
    args: [],
    nullable: true,
    defaultValue: 'bar',
    comment: undefined,
    priority: 0,
  } as Operation)
})

test('change anonymous index', async () => {
  const result = await computeDiff(dbFactory([
    tableFactory({
      name: 'User',
      columns: [
        columnFactory({
          name: 'id',
          type: 'increments',
        }),
        columnFactory({
          name: 'email',
          type: 'string',
        }),
      ],
      indexes: [{ columns: ['id'] }],
    }),
  ]), dbFactory([
    tableFactory({
      name: 'User',
      columns: [
        columnFactory({
          name: 'id',
          type: 'increments',
        }),
        columnFactory({
          name: 'email',
          type: 'string',
        }),
      ],
      indexes: [{ columns: ['email'] }],
    }),
  ]))
  expect(result.length).toEqual(2);
  expect(result[0]).toEqual({
    type: 'table.index.drop',
    table: 'User',
    indexName: undefined,
    columns: ['id'],
    priority: 0,
  } as Operation)
  expect(result[1]).toEqual({
    type: 'table.index.create',
    table: 'User',
    indexName: undefined,
    indexType: undefined,
    columns: ['email'],
    priority: 0,
  } as Operation)
})

test('change named index', async () => {
  const result = await computeDiff(dbFactory([
    tableFactory({
      name: 'User',
      columns: [
        columnFactory({
          name: 'id',
          type: 'increments',
        }),
        columnFactory({
          name: 'email',
          type: 'string',
        }),
      ],
      indexes: [{ columns: ['id'], name: 'foo' }],
    }),
  ]), dbFactory([
    tableFactory({
      name: 'User',
      columns: [
        columnFactory({
          name: 'id',
          type: 'increments',
        }),
        columnFactory({
          name: 'email',
          type: 'string',
        }),
      ],
      indexes: [{ columns: ['email'], name: 'foo' }],
    }),
  ]))
  expect(result.length).toEqual(2);
  expect(result[0]).toEqual({
    type: 'table.index.drop',
    table: 'User',
    columns: ['id'],
    indexName: 'foo',
    priority: 0,
  } as Operation)
  expect(result[1]).toEqual({
    type: 'table.index.create',
    table: 'User',
    columns: ['email'],
    indexName: 'foo',
    indexType: undefined,
    priority: 0,
  } as Operation)
})

test('untouched named index', async () => {
  const result = await computeDiff(dbFactory([
    tableFactory({
      name: 'User',
      columns: [
        columnFactory({
          name: 'id',
          type: 'increments',
        }),
        columnFactory({
          name: 'email',
          type: 'string',
        }),
      ],
      indexes: [{ columns: ['id'] }, { columns: ['id'], name: 'foo' }],
    }),
  ]), dbFactory([
    tableFactory({
      name: 'User',
      columns: [
        columnFactory({
          name: 'id',
          type: 'increments',
        }),
        columnFactory({
          name: 'email',
          type: 'string',
        }),
      ],
      indexes: [{ columns: ['id'], name: 'foo' }],
    }),
  ]))
  expect(result.length).toEqual(1);
  expect(result[0]).toEqual({
    type: 'table.index.drop',
    indexName: undefined,
    table: 'User',
    columns: ['id'],
    priority: 0,
  } as Operation)
})

test('create table & join table', async () => {
  const result = await computeDiff(dbFactory(), dbFactory([
    tableFactory({
      name: 'user',
    }),
    tableFactory({
      name: 'user_groups_join_group_users',
    }),
  ]))
  expect(result.length).toEqual(2);
  expect(result[0]).toEqual({
    type: 'table.create',
    table: 'user',
    priority: 0,
  } as Operation)
  expect(result[1]).toEqual({
    type: 'table.create',
    table: 'user_groups_join_group_users',
    priority: 1,
  } as Operation)
})
