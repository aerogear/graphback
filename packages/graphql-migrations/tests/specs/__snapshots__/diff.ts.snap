// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`create simple table 1`] = `
Array [
  Object {
    "priority": 0,
    "table": "User",
    "type": "table.create",
  },
  Object {
    "comment": "Some comment",
    "priority": 0,
    "table": "User",
    "type": "table.comment.set",
  },
  Object {
    "column": "id",
    "columnType": "increments",
    "priority": 0,
    "table": "User",
    "type": "table.primary.set",
  },
  Object {
    "args": Array [
      150,
    ],
    "column": "name",
    "columnType": "string",
    "comment": undefined,
    "defaultValue": undefined,
    "nullable": true,
    "priority": 0,
    "table": "User",
    "type": "column.create",
  },
]
`;
